### ISSUE TITLE
flycheck-emacs-lisp-initialize-packages auto does not work

### ISSUE THREAD

**Issue by stardiviner** [flycheck/flycheck#775]
I config Emacs flycheck with:
```elisp
(use-package flycheck
  :init
  ;; (add-hook 'after-init-hook #'global-flycheck-mode)
  (add-hook 'prog-mode-hook #'flycheck-mode)
  :config
  (setq flycheck-emacs-lisp-initialize-packages 'auto
        flycheck-emacs-lisp-load-path 'inherit
        ;; flycheck-emacs-lisp-package-user-dir nil
        )
)
```

It shows warning on variables with warining info `assignment to free variable ...`
Then I tried to debug with `emacs -q`, then load packages `package-initialize`. Then `global-flycheck-mode`, open emacs lisp file, still shows this warning.

**Comment by lunaryorn** [comment]
@stardiviner Please provide a minimal Emacs Lisp example that shows free variable warnings with this setup.

**Comment by stardiviner** [comment]
minimal elisp example:
filename: `~/.emacs.d/init/kk.el`
```lisp
(use-package flycheck
  :init
  ;; (add-hook 'after-init-hook #'global-flycheck-mode)
  (add-hook 'prog-mode-hook #'flycheck-mode)
  :config
  (setq flycheck-check-syntax-automatically '(save new-line)
        flycheck-idle-change-delay 5.0
        flycheck-display-errors-delay 0.9
        flycheck-highlighting-mode 'symbols
        flycheck-indication-mode 'left-fringe
        ;; 'flycheck-fringe-bitmap-double-arrow
        flycheck-standard-error-navigation t ; [M-g n/p]
        flycheck-deferred-syntax-check nil
        ;; flycheck-mode-line '(:eval (flycheck-mode-line-status-text))
        flycheck-completion-system nil ; 'ido, 'grizzl, nil
        )
  )
```

And here is a screenshot:
![flycheck emacs lisp load-path 1](https://cloud.githubusercontent.com/assets/378699/10866840/b70afefe-807e-11e5-979e-fd113b357ecb.png)

**Comment by stardiviner** [comment]
And here is a screenshot which run emacs with `emacs -q`, then `package-initialize`, then `global-flycheck-mode`.

http://imgur.com/09R5Se6

(PS. because github image upload failed many times, so I upload to other places.)

**Comment by stardiviner** [comment]
And other informations:
Emacs version: `24.5.1`
Linux system: `Arch latest`

**Comment by stardiviner** [comment]
There is other places which don't use `use-package` also have this warning.
```lisp
(unless (boundp 'my-inferior-shell-map)
  (define-prefix-command 'my-inferior-shell-map))
(define-key my-prog-inferior-map (kbd "s") 'my-inferior-shell-map)
```
Like this code snippet,
same environment with above.

**Comment by stardiviner** [comment]
Same warning also happens on this code snippet which don't use `use-package`.
```lisp
(unless (boundp 'my-inferior-shell-map)
  (define-prefix-command 'my-inferior-shell-map))
(define-key my-prog-inferior-map (kbd "s") 'my-inferior-shell-map)
```
Warning `reference to free variable my-prog-inferior-map`.
Same environment as above.

**Comment by lunaryorn** [comment]
@stardiviner Uhm, with all due respect, the warning is obviously correct from my point of view:  There's no `(defvar my-prog-inferior-map)` in this example, so the variable _is_ undefined.  What makes you think that this is a bug?!

**Comment by lunaryorn** [comment]
@stardiviner Type `C-c ! l` to show the list of errors in this buffer: You'll see that Flycheck also complains about `use-package` being undefined.

`use-package` is not autoloaded, so you need to add `(eval-when-compile (require 'use-package))` at the top of each file using this macro.  Otherwise Emacs (and thus Flycheck) doesn't know about the definition of the macro—it doesn't even know that it _is_ a macro.

Closing as `invalid`, because it's not an issue in Flycheck itself.

**Comment by stardiviner** [comment]
sorry, because of my CHina network issue, I failed many times on post on GitHub.
I defiend `my-prog-inferior-map` in other init files. All my init files are located in `~/.emacs.d/init/`.
I checked `load-path` of Emacs, it has `~/.emacs.d/init` path, weird, why flycheck does not found this definition.
I want to ask, if `use-packaged` add `###autoload`. Does it solve this problem?

**Comment by lunaryorn** [comment]
@stardiviner  Flycheck doesn't load all your init files automatically.  Hence it's not enough to define the variable in some other file.  You have to load the definition of the variable explicitly if it is defined in another file.  In other words you have to `require` the file that contains the definition if you use the variable in another file.   

Autoloading `use-package` would solve this issue.  Alternatively, add `(require 'use-package)` to your files.  I'd recommend the latter instead of waiting for a change in use-package.

**Comment by stardiviner** [comment]
Adding `(require 'use-package)` need to add many times. I will keep waiting.
And thanks for your help.

**Comment by jwiegley** [comment]
I'd strongly recommend adding this to your `use-package` declaration:

``` elisp
:commands flycheck-mode
```

This way it won't load `flycheck` at  startup, but only when a `prog-mode` has been entered.

**Comment by stardiviner** [comment]
@jwiegley Thanks
